### ISSUE TITLE
Dangling layers left behind after deleting image that was started with --uidmap

### ISSUE THREAD

**Issue by freva** [containers/podman#12316]
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind bug

**Description**

This issue was mentioned in https://github.com/containers/storage/issues/1060, but I figured it would be best with a separate issue.

After removing an image that was previously run by a container that was started with `--uidmap` only the top layer is removed. The remaining layers are left behind.

**Steps to reproduce the issue:**

1. (Optional) Reset podman
```
$ sudo podman system reset -f
A storage.conf file exists at /etc/containers/storage.conf
You should remove this file if you did not modified the configuration.
$ sudo ls -l /var/lib/containers/storage/
ls: cannot access '/var/lib/containers/storage/': No such file or directory
```
2. Pull an image with multiple layers:
```
$ sudo podman pull docker.io/library/node:16
...
15ddf4b49c29aae416981fe4bcfe06a3430ad5c509b1e7f4a0fd0135d13532f0
$ sudo du -skh /var/lib/containers/storage/
926M	/var/lib/containers/storage/
```
3. Start a container with image from previous step and `--uidmap`
```
$ sudo podman run --rm --uidmap 0:1:100000 docker.io/library/node:16
$ sudo du -skh /var/lib/containers/storage/
928M	/var/lib/containers/storage/
```
4. Remove the previous image: 
```
$ sudo podman rmi docker.io/library/node:16
Untagged: docker.io/library/node:16
Deleted: 15ddf4b49c29aae416981fe4bcfe06a3430ad5c509b1e7f4a0fd0135d13532f0
$ sudo du -skh /var/lib/containers/storage/
926M	/var/lib/containers/storage/
$ sudo cat /var/lib/containers/storage/overlay-images/images.json
[]
$ sudo cat /var/lib/containers/storage/overlay-layers/layers.json | jq '. | length'
8
```
If I pull the image again, it shows that all but the last layer are already there, only after I delete it again (without starting with `--uidmap`) will everything be removed.
```
$ sudo podman pull docker.io/library/node:16
Trying to pull docker.io/library/node:16...
Getting image source signatures
Copying blob 5402d6c1eb0a skipped: already exists  
Copying blob c6cef1aa2170 skipped: already exists  
Copying blob 13a51f13be8e skipped: already exists  
Copying blob def39d67a1a7 skipped: already exists  
Copying blob 07471e81507f skipped: already exists  
Copying blob a8367252e08e skipped: already exists  
Copying blob f06bf0d834a6 skipped: already exists  
Copying blob a05e3c4fa294 skipped: already exists  
Copying blob 01b4cbb6b46e done  
Copying config 15ddf4b49c done  
Writing manifest to image destination
Storing signatures
15ddf4b49c29aae416981fe4bcfe06a3430ad5c509b1e7f4a0fd0135d13532f0
$ sudo cat /var/lib/containers/storage/overlay-layers/layers.json | jq '. | length'
9
$ sudo podman rmi docker.io/library/node:16
Untagged: docker.io/library/node:16
Deleted: 15ddf4b49c29aae416981fe4bcfe06a3430ad5c509b1e7f4a0fd0135d13532f0
$ sudo cat /var/lib/containers/storage/overlay-layers/layers.json
[]
$ sudo du -skh /var/lib/containers/storage/
148K	/var/lib/containers/storage/
```

**Output of `podman info --debug`:**

```
host:
  arch: amd64
  buildahVersion: 1.21.3
  cgroupControllers:
  - cpuset
  - cpu
  - cpuacct
  - blkio
  - memory
  - devices
  - freezer
  - net_cls
  - perf_event
  - net_prio
  - hugetlb
  - pids
  - rdma
  cgroupManager: systemd
  cgroupVersion: v1
  conmon:
    package: conmon-2.0.29-1.module+el8.4.0+11822+6cc1e7d7.x86_64
    path: /usr/bin/conmon
    version: 'conmon version 2.0.29, commit: ae467a0c8001179d4d0adf4ada381108a893d7ec'
  cpus: 12
  distribution:
    distribution: '"rhel"'
    version: "8.4"
  eventLogger: file
  hostname: 68gj7d3
  idMappings:
    gidmap: null
    uidmap: null
  kernel: 4.18.0-305.19.1.el8_4.x86_64
  linkmode: dynamic
  memFree: 13493592064
  memTotal: 33244659712
  ociRuntime:
    name: runc
    package: runc-1.0.0-74.rc95.module+el8.4.0+11822+6cc1e7d7.x86_64
    path: /usr/bin/runc
    version: |-
      runc version spec: 1.0.2-dev
      go: go1.15.13
      libseccomp: 2.5.1
  os: linux
  remoteSocket:
    path: /run/podman/podman.sock
  security:
    apparmorEnabled: false
    capabilities: CAP_NET_RAW,CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_NET_BIND_SERVICE,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID,CAP_SYS_CHROOT
    rootless: false
    seccompEnabled: true
    seccompProfilePath: /usr/share/containers/seccomp.json
    selinuxEnabled: true
  serviceIsRemote: false
  slirp4netns:
    executable: ""
    package: ""
    version: ""
  swapFree: 50311983104
  swapTotal: 50377781248
  uptime: 200h 8m 23.29s (Approximately 8.33 days)
registries:
  search:
  - registry.access.redhat.com
  - registry.redhat.io
  - docker.io
store:
  configFile: /etc/containers/storage.conf
  containerStore:
    number: 0
    paused: 0
    running: 0
    stopped: 0
  graphDriverName: overlay
  graphOptions:
    overlay.mountopt: nodev,metacopy=on
  graphRoot: /var/lib/containers/storage
  graphStatus:
    Backing Filesystem: xfs
    Native Overlay Diff: "false"
    Supports d_type: "true"
    Using metacopy: "true"
  imageStore:
    number: 0
  runRoot: /run/containers/storage
  volumePath: /var/lib/containers/storage/volumes
version:
  APIVersion: 3.2.3
  Built: 1627570963
  BuiltTime: Thu Jul 29 17:02:43 2021
  GitCommit: ""
  GoVersion: go1.15.7
  OsArch: linux/amd64
  Version: 3.2.3
```

**Package info (e.g. output of `rpm -q podman` or `apt list podman`):**

```
podman-3.2.3-0.11.module+el8.4.0+12050+ef972f71.x86_64
```

**Have you tested with the latest version of Podman and have you checked the Podman Troubleshooting Guide? (https://github.com/containers/podman/blob/master/troubleshooting.md)**

Yes

**Additional environment details (AWS, VirtualBox, physical, etc.):**

**Comment by rhatdan** [comment]
This is the expected behavior since the underlying layers could be used by other images.  In this case the UID mapping is actually creating a separate image. What would you expect the command to do. @nalind @giuseppe WDYT
We could check to see if we are removing a UID mapped image, to remove the underlying storage if no other images are using it.
When we move to idmapped images, this will change the current behaviour and could cause unexpected behavior.

**Comment by freva** [comment]
Maybe I'm missing something, but how am I supposed to remove these? `podman images` doesn't show anything, `podman image prune` doesn't remove anything, so this just uses up disk space.

**Comment by rhatdan** [comment]
If you do `podman rmi IMAGE` twice it removes it correct?

**Comment by freva** [comment]
If I re-pull the image in between, yes. I cant run the rmi command twice in a row (fails with "image not known").

**Comment by rhatdan** [comment]
I am saying does this work and clear out the image?
```
$ sudo podman pull docker.io/library/node:16
$ sudo podman run --rm --uidmap 0:1:100000 docker.io/library/node:16
$ sudo podman rmi docker.io/library/node:16
$ sudo podman rmi docker.io/library/node:16
```

**Comment by freva** [comment]
Yeah, I got it, the rmi command doesnt work twice in a row, the second one fails with "image not known":
```
$ sudo podman system reset -f
A storage.conf file exists at /etc/containers/storage.conf
You should remove this file if you did not modified the configuration.
$ sudo podman pull docker.io/library/node:16
...
15ddf4b49c29aae416981fe4bcfe06a3430ad5c509b1e7f4a0fd0135d13532f0
$ sudo podman run --rm --uidmap 0:1:100000 docker.io/library/node:16
$ sudo podman rmi docker.io/library/node:16 
Untagged: docker.io/library/node:16
Deleted: 15ddf4b49c29aae416981fe4bcfe06a3430ad5c509b1e7f4a0fd0135d13532f0
$ sudo podman rmi docker.io/library/node:16
Error: docker.io/library/node:16: image not known
$ sudo du -skh /var/lib/containers/storage/
926M	/var/lib/containers/storage/
```

**Comment by rhatdan** [comment]
What does podman images show?

**Comment by freva** [comment]
No images

**Comment by rhatdan** [comment]
No images, but you still see storage?

**Comment by rhatdan** [comment]
@nalind PTAL

**Comment by freva** [comment]
Yes
```
$ sudo find /var/lib/containers/storage/ -maxdepth 2
/var/lib/containers/storage/
/var/lib/containers/storage/libpod
/var/lib/containers/storage/libpod/bolt_state.db
/var/lib/containers/storage/mounts
/var/lib/containers/storage/tmp
/var/lib/containers/storage/overlay
/var/lib/containers/storage/overlay/l
/var/lib/containers/storage/overlay/d6a325d281f2a509f28cb5f48f80d48f979b4b60873f9c42724c3ba3ccd90c4c
/var/lib/containers/storage/overlay/8932e076b1141abeda3580cc6a4ba7fad62006609725319aa1a698a6491d5f69
/var/lib/containers/storage/overlay/4ddc486ef1c3d5912ae188bfc6332a872b105142836dd42ef12ced5b0a4f5f79
/var/lib/containers/storage/overlay/ce5b0a73f5116c79617fe181afae83934ad7ca15963e52fbbeb648751acb4a2a
/var/lib/containers/storage/overlay/788eae02005ef36fe2e41760f62682471d72329aab8beabd3940f40330eb0515
/var/lib/containers/storage/overlay/aa8a03cd911297c99efead8d0e2db80394aea89ce405a7d596e789d76cc3ba9b
/var/lib/containers/storage/overlay/1f39d72e370159465bc55554db87cd738ebd1da61764ebc9fcad0f45205bde6b
/var/lib/containers/storage/overlay/7005f61eb112ff344487992abe0d4ac9ba44bcf12fa3ee4dcb2582fe667079e6
/var/lib/containers/storage/overlay/backingFsBlockDev
/var/lib/containers/storage/storage.lock
/var/lib/containers/storage/userns.lock
/var/lib/containers/storage/overlay-images
/var/lib/containers/storage/overlay-images/images.lock
/var/lib/containers/storage/overlay-images/images.json
/var/lib/containers/storage/overlay-containers
/var/lib/containers/storage/overlay-containers/containers.lock
/var/lib/containers/storage/overlay-containers/containers.json
/var/lib/containers/storage/cache
/var/lib/containers/storage/cache/blob-info-cache-v1.boltdb
/var/lib/containers/storage/overlay-layers
/var/lib/containers/storage/overlay-layers/layers.lock
/var/lib/containers/storage/overlay-layers/layers.json
/var/lib/containers/storage/overlay-layers/d6a325d281f2a509f28cb5f48f80d48f979b4b60873f9c42724c3ba3ccd90c4c.tar-split.gz
/var/lib/containers/storage/overlay-layers/8932e076b1141abeda3580cc6a4ba7fad62006609725319aa1a698a6491d5f69.tar-split.gz
/var/lib/containers/storage/overlay-layers/4ddc486ef1c3d5912ae188bfc6332a872b105142836dd42ef12ced5b0a4f5f79.tar-split.gz
/var/lib/containers/storage/overlay-layers/ce5b0a73f5116c79617fe181afae83934ad7ca15963e52fbbeb648751acb4a2a.tar-split.gz
/var/lib/containers/storage/overlay-layers/788eae02005ef36fe2e41760f62682471d72329aab8beabd3940f40330eb0515.tar-split.gz
/var/lib/containers/storage/overlay-layers/aa8a03cd911297c99efead8d0e2db80394aea89ce405a7d596e789d76cc3ba9b.tar-split.gz
/var/lib/containers/storage/overlay-layers/1f39d72e370159465bc55554db87cd738ebd1da61764ebc9fcad0f45205bde6b.tar-split.gz
/var/lib/containers/storage/overlay-layers/7005f61eb112ff344487992abe0d4ac9ba44bcf12fa3ee4dcb2582fe667079e6.tar-split.gz
```
The `images.json` is empty, but `layers.json` has content (see original post). All the IDs under overlay are also in overlay-layers.

**Comment by hakonhall** [comment]
Which podman version will this be included in?

https://github.com/containers/storage has not yet made any releases since September 30, and AFAIK Podman does not mention this issue as included in the latest 3.4.x releases.

**Comment by mheon** [comment]
Usually we only do significant version bumps to dependencies in new minor releases, so it would have to wait until Podman 4.0.0 in late January. However, we had to pick up a new c/image in 3.4.4 that also brought a new c/storage, so I think it may be in the v3.4.4 release?
